File: scripts/indexnow-bulk-submit.py
#!/usr/bin/env python3

"""
IndexNow Bulk Submission Script (Python)
Submits all URLs from Hugo sitemap to IndexNow API
"""

import json
import urllib.request
import urllib.parse
import xml.etree.ElementTree as ET
import time
import os
import sys

# Configuration
CONFIG = {
    'host': 'pablobernabeu.github.io',
    'api_key': 'ba7d2697a8f44966bd90543d188a8aac',
    'key_location': 'https://pablobernabeu.github.io/ba7d2697a8f44966bd90543d188a8aac.txt',
    'sitemap_path': 'public/sitemap.xml',
    'api_endpoint': 'https://api.indexnow.org/indexnow',
    'batch_size': 50,  # Smaller batches to avoid 403 errors
    'delay_between_batches': 3  # Longer delay between batches
}

def extract_urls_from_sitemap(sitemap_path):
    """Extract URLs from Hugo sitemap.xml"""
    try:
        tree = ET.parse(sitemap_path)
        root = tree.getroot()
        
        urls = []
        # Handle namespace
        namespace = {'ns': 'http://www.sitemaps.org/schemas/sitemap/0.9'}
        
        for url in root.findall('ns:url', namespace):
            loc = url.find('ns:loc', namespace)
            if loc is not None:
                urls.append(loc.text)
        
        # If no namespace, try without
        if not urls:
            for url in root.findall('url'):
                loc = url.find('loc')
                if loc is not None:
                    urls.append(loc.text)
        
        return urls
    except Exception as e:
        print(f"Error reading sitemap: {e}")
        return []

def submit_to_indexnow(urls):
    """Submit URLs to IndexNow API"""
    payload = {
        'host': CONFIG['host'],
        'key': CONFIG['api_key'],
        'keyLocation': CONFIG['key_location'],
        'urlList': urls
    }
    
    try:
        data = json.dumps(payload).encode('utf-8')
        
        req = urllib.request.Request(
            CONFIG['api_endpoint'],
            data=data,
            headers={
                'Content-Type': 'application/json; charset=utf-8',
                'User-Agent': 'Hugo-IndexNow-Submitter/1.0'
            },
            method='POST'
        )
        
        with urllib.request.urlopen(req) as response:
            return {
                'success': True,
                'status': response.getcode(),
                'response': response.read().decode('utf-8')
            }
    except urllib.error.HTTPError as e:
        return {
            'success': e.code in [200, 202],
            'status': e.code,
            'error': str(e)
        }
    except Exception as e:
        return {
            'success': False,
            'error': str(e)
        }

def submit_in_batches(urls):
    """Split URLs into batches and submit"""
    batches = [urls[i:i + CONFIG['batch_size']] for i in range(0, len(urls), CONFIG['batch_size'])]
    
    print(f"📊 Submitting {len(urls)} URLs in {len(batches)} batch(es)")
    
    for i, batch in enumerate(batches, 1):
        print(f"\n🚀 Submitting batch {i}/{len(batches)} ({len(batch)} URLs)...")
        
        result = submit_to_indexnow(batch)
        
        if result['success']:
            print(f"✅ Batch {i} submitted successfully (Status: {result['status']})")
        else:
            print(f"❌ Batch {i} failed (Status: {result.get('status', 'Error')}): {result.get('error', 'Unknown error')}")
        
        # Rate limiting delay
        if i < len(batches):
            print(f"⏳ Waiting {CONFIG['delay_between_batches']}s before next batch...")
            time.sleep(CONFIG['delay_between_batches'])

def main():
    print('🔍 IndexNow Bulk Submission Script (Python)')
    print('==========================================')
    
    # Check if sitemap exists
    if not os.path.exists(CONFIG['sitemap_path']):
        print(f"❌ Sitemap not found at: {CONFIG['sitemap_path']}")
        print('💡 Make sure to build your Hugo site first with: hugo')
        sys.exit(1)
    
    # Extract URLs from sitemap
    print(f"📄 Reading sitemap from: {CONFIG['sitemap_path']}")
    urls = extract_urls_from_sitemap(CONFIG['sitemap_path'])
    
    if not urls:
        print('⚠️  No URLs found in sitemap')
        sys.exit(0)
    
    print(f"📊 Found {len(urls)} URLs in sitemap")
    
    # Filter to only include your domain
    filtered_urls = [url for url in urls if CONFIG['host'] in url]
    print(f"🎯 Filtered to {len(filtered_urls)} URLs from your domain")
    
    if not filtered_urls:
        print('⚠️  No URLs found for your domain')
        sys.exit(0)
    
    # Submit URLs
    submit_in_batches(filtered_urls)
    
    print('\n🎉 Bulk submission completed!')
    print('\n📝 Next steps:')
    print('1. Check search console for indexing status')
    print('2. New pages will be automatically submitted via JavaScript')
    print('3. Use GitHub Actions for automatic submission on deployment')

if __name__ == '__main__':
    main()

File: scripts/indexnow-bulk-submit.js
#!/usr/bin/env node

/**
 * IndexNow Bulk Submission Script
 * Submits all URLs from Hugo sitemap to IndexNow API
 */

const fs = require("fs");
const path = require("path");
const fetch = require("node-fetch");
const xml2js = require("xml2js");

// Configuration
const CONFIG = {
  host: "pablobernabeu.github.io",
  apiKey: "ba7d2697a8f44966bd90543d188a8aac",
  keyLocation:
    "https://pablobernabeu.github.io/ba7d2697a8f44966bd90543d188a8aac.txt",
  sitemapPath: path.join(__dirname, "..", "public", "sitemap.xml"),
  apiEndpoint: "https://api.indexnow.org/indexnow",
  batchSize: 1000, // Maximum URLs per submission (Bing recommends ≤10,000)
  delayBetweenBatches: 2000, // 2 seconds between batches
};

/**
 * Extract URLs from Hugo sitemap.xml
 */
async function extractUrlsFromSitemap(sitemapPath) {
  try {
    const sitemapContent = fs.readFileSync(sitemapPath, "utf8");
    const parser = new xml2js.Parser();
    const result = await parser.parseStringPromise(sitemapContent);

    const urls = [];
    if (result.urlset && result.urlset.url) {
      for (const url of result.urlset.url) {
        if (url.loc && url.loc[0]) {
          urls.push(url.loc[0]);
        }
      }
    }

    return urls;
  } catch (error) {
    console.error("Error reading sitemap:", error.message);
    return [];
  }
}

/**
 * Submit URLs to IndexNow API
 */
async function submitToIndexNow(urls) {
  const payload = {
    host: CONFIG.host,
    key: CONFIG.apiKey,
    keyLocation: CONFIG.keyLocation,
    urlList: urls,
  };

  try {
    const response = await fetch(CONFIG.apiEndpoint, {
      method: "POST",
      headers: {
        "Content-Type": "application/json; charset=utf-8",
      },
      body: JSON.stringify(payload),
    });

    return {
      success: response.ok || response.status === 202,
      status: response.status,
      statusText: response.statusText,
    };
  } catch (error) {
    return {
      success: false,
      error: error.message,
    };
  }
}

/**
 * Split URLs into batches and submit
 */
async function submitInBatches(urls) {
  const batches = [];
  for (let i = 0; i < urls.length; i += CONFIG.batchSize) {
    batches.push(urls.slice(i, i + CONFIG.batchSize));
  }

  console.log(
    `📊 Submitting ${urls.length} URLs in ${batches.length} batch(es)`
  );

  for (let i = 0; i < batches.length; i++) {
    const batch = batches[i];
    console.log(
      `\n🚀 Submitting batch ${i + 1}/${batches.length} (${
        batch.length
      } URLs)...`
    );

    const result = await submitToIndexNow(batch);

    if (result.success) {
      console.log(
        `✅ Batch ${i + 1} submitted successfully (Status: ${result.status})`
      );
    } else {
      console.log(
        `❌ Batch ${i + 1} failed (Status: ${result.status || "Error"}): ${
          result.statusText || result.error
        }`
      );
    }

    // Rate limiting delay between batches
    if (i < batches.length - 1) {
      console.log(
        `⏳ Waiting ${CONFIG.delayBetweenBatches}ms before next batch...`
      );
      await new Promise((resolve) =>
        setTimeout(resolve, CONFIG.delayBetweenBatches)
      );
    }
  }
}

/**
 * Main execution
 */
async function main() {
  console.log("🔍 IndexNow Bulk Submission Script");
  console.log("=====================================");

  // Check if sitemap exists
  if (!fs.existsSync(CONFIG.sitemapPath)) {
    console.error(`❌ Sitemap not found at: ${CONFIG.sitemapPath}`);
    console.log("💡 Make sure to build your Hugo site first with: hugo");
    process.exit(1);
  }

  // Extract URLs from sitemap
  console.log(`📄 Reading sitemap from: ${CONFIG.sitemapPath}`);
  const urls = await extractUrlsFromSitemap(CONFIG.sitemapPath);

  if (urls.length === 0) {
    console.log("⚠️  No URLs found in sitemap");
    process.exit(0);
  }

  console.log(`📊 Found ${urls.length} URLs in sitemap`);

  // Filter to only include your domain
  const filteredUrls = urls.filter((url) => url.includes(CONFIG.host));
  console.log(`🎯 Filtered to ${filteredUrls.length} URLs from your domain`);

  if (filteredUrls.length === 0) {
    console.log("⚠️  No URLs found for your domain");
    process.exit(0);
  }

  // Submit URLs
  await submitInBatches(filteredUrls);

  console.log("\n🎉 Bulk submission completed!");
  console.log("\n📝 Next steps:");
  console.log("1. Check search console for indexing status");
  console.log("2. New pages will be automatically submitted via JavaScript");
  console.log("3. Use GitHub Actions for automatic submission on deployment");
}

// Run the script
main().catch((error) => {
  console.error("💥 Script failed:", error.message);
  process.exit(1);
});
